support-frontend/app/services/MParticleClient.scala

```diff
@@ -0,0 +1,188 @@
+)
+
+case class ProfileRequest(
+    environment_type: String,
+    identity: Identity,
+)
+
+case class AudienceMembership(
+    audience_id: Int,
+)
+
+case class ProfileResponse(
+    audience_memberships: List[AudienceMembership],
+)
+
+object MParticleClient {
+  implicit val oauthTokenRequestEncoder: Encoder[OAuthTokenRequest] = deriveEncoder
+  implicit val oauthTokenResponseDecoder: Decoder[OAuthTokenResponse] = deriveDecoder
+  implicit val identityEncoder: Encoder[Identity] = deriveEncoder
+  implicit val profileRequestEncoder: Encoder[ProfileRequest] = deriveEncoder
+  implicit val audienceMembershipDecoder: Decoder[AudienceMembership] = deriveDecoder
+  implicit val profileResponseDecoder: Decoder[ProfileResponse] = deriveDecoder
+}
+
+class MParticleClient(
+    wsClient: WSClient,
+    mparticleConfigProvider: MparticleConfigProvider,
+)(implicit ec: ExecutionContext)
+    extends SafeLogging {
+
+  private lazy val mparticleConfig: MparticleConfig = mparticleConfigProvider.get()
+
+  private case class CachedToken(token: String, expiresAt: DateTime)
+  private val tokenCache = new AtomicReference[Option[CachedToken]](None)
+  private val safetyMargin = 2.minutes
+
+  def getUserProfile(identityId: String): Future[Option[MParticleUserProfile]] = {
+    getCachedAccessToken()
+      .flatMap { accessToken =>
+        callProfileAPI(identityId, accessToken.token).map(responseBody => Some(parseUserProfile(responseBody)))
+      }
+      .recover { case ex =>
+        logger.error(scrub"Failed to get mParticle user profile: ${ex.getMessage}", ex)
+        None
+      }
+  }
+
+  private def safeBodyHash(body: String): String = {
+    val hash = body.hashCode.abs
+    s"hash=$hash"
+  }
+
+  private def getCachedAccessToken(): Future[MParticleAccessToken] = {
+    val now = DateTime.now()
+
+    tokenCache.get() match {
+      case Some(cachedToken) if cachedToken.expiresAt.isAfter(now) =>
+        Future.successful(MParticleAccessToken(cachedToken.token))
+      case _ =>
+        getAccessToken().map { case (token, expiresInSeconds) =>
+          val actualExpiryTime = now.plusSeconds(expiresInSeconds)
+          val safeExpiryTime = actualExpiryTime.minus(safetyMargin.toMillis)
+          tokenCache.set(Some(CachedToken(token, safeExpiryTime)))
+          logger.info(
+            s"Cached new mParticle access token, expires at: $safeExpiryTime (${safetyMargin.toSeconds}s safety margin)",
+          )
+          MParticleAccessToken(token)
+        }
+    }
+  }
+
+  private def getAccessToken(): Future[(String, Int)] = {
+    import MParticleClient._
+
+    val request = OAuthTokenRequest(
+      client_id = mparticleConfig.clientId,
+      client_secret = mparticleConfig.clientSecret,
+      audience = mparticleConfig.apiUrl,
+      grant_type = "client_credentials",
+    )
+
+    wsClient
+      .url(mparticleConfig.tokenUrl)
+      .withHttpHeaders("Content-Type" -> "application/json")
+      .post(request.asJson.noSpaces)
```

Comment:
do the other Clients in this repo use wsClient directly, or is there some kind of RESTClient that you can reuse?